### PR TITLE
Issues/login jetpack discovery

### PR DESCRIPTION
**Pull request by aerych**
This PR adds a check for Jetpack to the end of the XML-RPC discovery process.  If we're unable to find the XML-RPC endpoint of a Jetpack connected blog, this look up will allow us to direct the user to Jetpack support to resolve the issue.
Note that this PR leaves intact the error override in `[WordPressXMLRPCAPIFacade guessXMLRPCURLForSite:success:failure]` and error handling will be updated in a subsequent PR.

To test:
You'll want a self-hosted blog to test with. 
You'll want to add a breakpoint in either `WordPressXMLRPCAPIFacade` before the error is over written, or in `WordPressOrgXMLRPCValidator` so you can inspect the error returned.

Scenario 0: 
Make sure you can login to your self-hosted site.

Scenario 1:
Prep: Edit your xmlrpc.php so it throws a PHP error.
Attempt to sign into your self-hosted site. 
Confirm the error indicates Jetpack is inaccessable.

Scenario 2:
Prep:
Edit your .htaccess to restrict access to xmlrpc.php.
Attempt to sign into your self-hosted site. 
Confirm the error indicates Jetpack is inaccessable.

Scenario 3: 
Prep: Install Jetpack but ensure it is disconnected.
Attempt to sign into your self-hosted site.
Confirm the error indicates Jetpack is disconnected.

Needs review: @koke would you do the honors?

cc @nheagy

### CODE REVIEW

**Comment by koke** [comment]
I'm not sure I understand what we're trying to solve here. If xmlrpc.php isn't accessible, won't Jetpack also be broken? Can you give me a bit more context?

**Comment by aerych** [comment]
> If xmlrpc.php isn't accessible, won't Jetpack also be broken? Can you give me a bit more context?

Totally broken, yep.   The goal here is to learn if a site is a Jetpack site. If it is we will provide tailored feedback, perhaps directing a user to Jetpack support rather than just saying "No WordPress site at that address."

File: WordPress/Classes/Networking/JetpackServiceRemote.m [hunk 1]
```diff
@@ -59,4 +60,75 @@ - (void)validateJetpackUsername:(NSString *)username
...
+                                                                  code:JetpackServiceRemoteErrorSiteInaccessable
+                                                              userInfo:error.userInfo];
+
+                                  }
+
+                                  // The unauthorized key is a special case here.
+                                  // The API call does not require credentials and does not return a 403 itself.
+                                  // This combination occurs when Jetpack is installed but not connected to wpcom.
+                                  if ([errorKey isEqualToString:@"unauthorized"] && error.code == WordPressComRestApiErrorAuthorizationRequired) {
+                                      // Jetpack is disabled/installed but not connected. (403 error)
```
Comment (koke) [inline WordPress/Classes/Networking/JetpackServiceRemote.m]: This error will be present if Jetpack was ever activated and a connection initiated, even after deleting Jetpack.

For a new site, the API returns:

- Before Jetpack: 404 unknown_blog
- Install Jetpack: 404 unknown_blog
- Activate Jetpack: 404 unknown_blog
- Start connecting: 403 unauthorized
- Deactivate Jetpack: 403 unauthorized
- Delete Jetpack: : 403 unauthorized

**Comment by koke** [comment]
To summarize, there are a couple of tricky cases:

- The site previously had Jetpack, and wp.com still thinks that it has
- The site has Jetpack installed/activated but hasn't been connected yet.

If you look at the wizard at https://wordpress.com/jetpack/connect/install it can successfully detect if jetpack is missing/installed/activated, so we might want to leverage the same backend.

**Comment by aerych** [comment]
I didn't know about `/jetpack/connect/install`, thanks for that tip!  I took a look and see that it calls an endpoint I was unaware of: `/connect/site-info/$protocol/$address` which does *EXACTLY* what this PR was attempting, only better and cleaner.  I'll update to call that endpoint instead.  Thanks again for sharing that find! :)

**Comment by aerych** [comment]
I've updated this branch to call `/connect/site-info/$protocol/$address` and check the result to see if the URL in question belongs to a Jetpack site.  For now this is indicated by the presence of a key in the error's user info dictionary (final error handling still to be done in a subsequent PR).

Ready for another round @koke

File: WordPress/Classes/Networking/JetpackServiceRemote.m [hunk 2]
```diff
@@ -59,4 +60,31 @@ - (void)validateJetpackUsername:(NSString *)username
          }];
 }
 
-@end
\ No newline at end of file
+/**
+    Check if the specified site is a Jetpack site.  The success block
+    receives a bool indicating if the site hass Jetpack.
```
Comment (koke) [inline WordPress/Classes/Networking/JetpackServiceRemote.m]: typo: hass ;)

File: WordPress/Classes/Networking/JetpackServiceRemote.m [hunk 3]
```diff
@@ -59,4 +60,31 @@ - (void)validateJetpackUsername:(NSString *)username
...
+ */
+- (void)checkSiteHasJetpack:(NSURL *)siteURL
+                    success:(void (^)(BOOL isJetpack))success
+                    failure:(void (^)(NSError *error))failure
+{
+    NSString *siteStr = [[NSString stringWithFormat:@"%@%@", siteURL.host, siteURL.path]
+                         stringByAddingPercentEncodingWithAllowedCharacters:NSCharacterSet.URLHostAllowedCharacterSet];
```
Comment (koke) [inline WordPress/Classes/Networking/JetpackServiceRemote.m]: I think this should use `URLPathAllowedCharacterSet` instead

Comment (aerych) [inline WordPress/Classes/Networking/JetpackServiceRemote.m]: I thought so too initially but it allows a `/` character which if included in the blog's path will cause the API call to resolve to a 404.  Try it with a self-hosted site installed to a subfolder to reproduce.  We need to escape any `\` characters in the `$address` component of the API call.  Hence using `.URLHostAllowedCharacterSet`.

**Comment by aerych** [comment]
Thank you @koke!!

**Pull request merged**